### DISCUSSION TITLE
Post-processing of Peridigm results

### DISCUSSION THREAD

**Discussion by jritta** [peridigm/peridigm#169] [Q&A] [answered]
Dear community,

How do you process the Exodus results provided by Peridigm? Is there a smart way to do it?

I found on my search on the web the following:
* @djlittl mentions the possibility of process the data by the `exdous.py` wrapper (#103, #91)
* @johntfoster mentions in his [blog](https://johnfoster.pge.utexas.edu/blog/posts/extracting-exodus-information-with-netcdf-python/) to access the data by netcdf library 
* In [Peridigm User Guid by DLR](https://elib.dlr.de/126320/1/Peridigm_Users_Guide.pdf) and #103 processing with ParaView is show.
* Access it by the [python library ](https://salvushub.github.io/pyexodus/)`pyexdous`

I am searching for a possibility to handle the Exodus results in a smart way since always using ParaViews GUI or hard code the data extraction via `exodus.py` is not very comfortable. Furthermore, and this is my main issue, I would like to edit the results file, e.g. reduce the size of the exodus file by deleting every second time step, cut the time series into chunks or crop it, compute time derivatives of some quantities, and so on.
It would be nice to have some unified, easy to use and well documented interface or accessor functions for the exodus result files.

Best
Jonas

**Comment by johntfoster** [comment]
I have an [older blog post](https://johnfoster.pge.utexas.edu/blog/posts/using-exoduspy-to-extract-data-from-fea-database/) that shows a simple example of how to use `exodus.py`.  To reduce the number of steps in a file, you can use the Trilinos utility `epu`, e.g.

`epu -steps 1:LAST:2 my_exodus_file`

should take every other time step.  You can also specify only the variables you're interested in with the `nvar` flag.  See the epu help with `epu -h` for more information.

**Comment by djlittl** [comment, marked as answer]
The SEACAS tool suite has many tools for operating on exodus files (epu is one of those tools, as John mentions above):

https://github.com/gsjaardema/seacas

Note that most of SEACAS is probably already included in your Trilinos build.

One other point is that exodus is built on netcdf.  You can read the files directly into Matlab using the Matlab ncread functions.  This is not super user friendly, but if you're a Matlab user you might find this useful.

**Comment by jritta** [comment]
@johntfoster and @djlittl thank you very much for your answers!

I had a closer look to the tools but I still facing a problem I did not solve yet. Maybe you can give me an hint or you have an idea how I could implement this. I want to compute a rate field variable of the data. That means I want to subtract a field variable of time step `n` from a field variable `n+1` and write it for each time step into the exodus file or to a new exodus file. Do you maybe have an idea how I could implement that (I would prefer a python implementation?

My current implementation is with `exomerge`:
```python
import exomerge3
import numpy as np

model = exomerge3.import_model('PATH_TO_EXODUS_FILE')
count = 0
for time in model.get_timesteps():
    values = np.array(model.get_node_field_values('NodeFieldVariable', time))
    if count == 0:
        previousValues = np.zeros_like(values)
        previousTime = 0.0
        count += 1
    rateValue = (previousValues - values)/(previousTime - time)
    previousValues = values

    model.create_node_field('Rate_Value', value=list(rateValue)[0])
model.export_model('PATH_TO_OUTPUT_FILE')
```
The problem here is the writing of the computed field to the exodus file.

1. With current implementation I override in each loop cycle the previous value (I can not specify the time step).
2. The method allows only to write one value for the whole field instead of an array such that I have for each point a different value.

I would appreciate some help in this regard! 

Best,
Jonas

**Comment by Martin1330** [comment]
我已收到您的邮件，祝您工作愉快！
     张以恒

**Comment by jritta** [comment]
Unfortunately, the epu command to reduce the number of time steps in an Exodus database did not work for me, but with the python interface exomerge I was able to write a script to do it. Nevertheless, for large databases this is not ideal because the library loads at the beginning the whole model to the memory and thus one is limited by the available computer memory.

Best,
Jonas
